spec/Candid.md

```diff
@@ -882,200 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
-
--------------------
-true ~> true : bool
-
----------------------
-false ~> false : bool
-
------------------------
-<text> ~> <text> : text
-
--------------------
-null ~> null : null 
+C[<t> <: <t>](x) = x    for every <t> ∈ <numtype>, bool, text, null
 ```
 
-Values of type `nat` coerce successfully at type `int`:
+Values of type `nat` coerce at type `int`:
 ```
---------------------------------
-(<x> : nat) ~> (<x> : int) : int
+C[nat <: int](<nat>) = <nat>
 ```
 
-Any value coerces at type `reserved`, producing the canonical value of type `reserved`:
-
+Coercion into `reserved` is the constant map (this is arbitrarily using `null` as “the” value of `reserved`):
 ```
------------------------------------
-<v> ~> (null : reserved) : reserved
+C[<t> <: reserved](_) = null
 ```
-
-NB: No rule is needed for type `empty`, because there are no values of that type. By construction, `<v> ~> _ : empty` will not hold.
+NB: No rule is needed for type `empty`, because there are no values of that type. By construction, `C[empty <: <t>]` is the unique function on the empty domain.
 
 #### Vectors
 
-Only vectors coerce at vector types, and only if all elements coerce successfully.
-
+Vectors coerce pointwise:
 ```
-<v> ~> <v'> : <t>
------------------------------------------
-vec { <v>;* } ~> vec { <v'>;* } : vec <t>
+C[vec <t> <: vec <t'>]( vec { <v>;* } ) = vec { C[<t> <: <t'>](<v>);* }
 ```
 
 #### Options
 
-The null value coerce at any option type:
+The null type coerces into any option type:
 ```
-----------------------
-null ~> null : opt <t>
+C[null <: opt <t>](null) = null
 ```
 
-An optional value coerces at an option type, if the constituent value coerces at the constituent type:
-```
-<v> ~> <v'> : <t>
------------------------------
-opt <v> ~> opt <v'> : opt <t>
+An optional value coerces at an option type, if the constituent value has a suitable type, and else goes to `null`:
 ```
-
-If an optional value _fails_ to coerce at an optional type, or the value is `reserved`, the result is `null`, not failure:
+C[opt <t> <: opt <t'>](null) = null
+C[opt <t> <: opt <t'>](opt <v>) = opt C[<t> <: <t'>](v)  if <t> <: <t'>
+C[opt <t> <: opt <t'>](opt <v>) = null                   if not(<t> <: <t'>)
 ```
-not (<v> ~> _ : <t>)
--------------------------
-opt <v> ~> null : opt <t>
 
------------------------------------
-(null : reserved) ~> null : opt <t>
-```
+NOTE: These rules above imply that a Candid decoder has to be able to decide the subtyping relation at runtime.
 
 Coercing a non-null, non-optional and non-reserved value at an option type treats it as an optional value:
 ```
-<v> ≠ null
-<v> ≠ (null : reserved)
-<v> ≠ opt _
-not (null <: <t>)
-opt <v> ~> <v'> : opt <t>
--------------------------
-<v> ~> <v'> : opt <t>
+C[<t> <: opt <t'>](<v>) = opt C[<t> <: <t'>](v)  if not (null <: <t>) and <t> <: <t'>
+C[<t> <: opt <t'>](_)   = null                   if not (null <: <t>) and not (<t> <: <t'>)
```

Comment:
```suggestion
C[<t> <: opt <t'>](<v>) = opt C[<t> <: <t'>](v)  if not (null <: <t'>) and <t> <: <t'>
C[<t> <: opt <t'>](_)   = null                   if not (null <: <t'>) and not (<t> <: <t'>)
```

These conditions need to be stricter, see https://dfinity.slack.com/archives/C016H8MCVRC/p1617755023039700

Reply:
How about the empty record case?
`C[ μx. record { x } <: μx. opt record { x } ](_) = opt C[μx. record { x } <: record { μx. opt record { x } } ]`

Do we want to allow `μx. record { x }` to be subtype of `opt t`?

Reply:
Every type is a subtype of `opt t`. The question is only whether it coerces to `null` or something else.

I think this case is ok: Note that the coercion function takes a value `v : µx.record{x}`, and no such value exists, so `C[ μx. record { x } <: μx. opt record { x } ]` is (independent of the equations) the empty function with an empty domain.

Reply:
I'm not sure how to detect loops in this case. The usual trick is to count the depth of the continuous nesting of records. But here, `µx.record{x}` doesn't change, we are only reducing the expected type, i.e. `C[ μx. record { x } <: μx. opt record { x } ] = opt C[ μx. record { x } <: μx. opt record { x } ]`. Operationally, the type on the left doesn't change at all, so we cannot count the depth. 

Of course, there are more expensive ways to detect loops, e.g. memoization or decoding the value with the actual type first. But to rule out other possible implementations just because of this case seems a bit wasteful.

Reply:
Is this really a new problem? The problem with `µx.record{x}` is about decoding, not coercion, and that isn’t changed by this PR? There are no values of that type anyways, so one wouldn’t ever get past the decoding stage before erroring out (or looping.)

In the Motoko decoder, I bump a counter whenever I go from `record { t1, … }` to `t1`, and reset whenever I actually consum a byte from the input stream. If that counter exceeds the table size I have recognized a loop.

In the Haskell decoder, I pre-process the type table and replace any empty type like `µx. record { …., x , …}` with `empty`.